File: p5/solution_code/C++/Queue.cpp
#include <tgmath.h>
#include "Stack.cpp"

template <typename Item>
class Queue {
    private:
	    Stack<Stack<Item>> enqueueingStacks;
	    Stack<Stack<Item>> dequeueingStacks;

	    Stack<Item> flipTo;
	    Stack<Item> flipFrom;
	    Stack<Item> enqueueTo;

	    Stack<Item> dequeueFrom;
	    Stack<Item> dequeueFromLast;

	    int stackExponent = 0;
        bool dequeueing = false;

    public:
        void enqueue(Item item) {
            if (flipFrom.size() > 0) {
                flipTo.push(flipFrom.pop());
            }

            enqueueTo.push(item);

            if (enqueueTo.size() == pow(2, stackExponent)) {
                enqueueingStacks.push(flipTo);
                flipTo = flipFrom;
                flipFrom = enqueueTo;
                enqueueTo = Stack<Item>();
                stackExponent++;
            }
        }

        Item dequeue() {
            for (int i = 0; i < 4; i++) {
                if (enqueueTo.size() > 0) {
                    dequeueFromLast.push(enqueueTo.pop());
                } else if (flipTo.size() > 0) {
                    flipFrom.push(flipTo.pop());
                } else if (flipFrom.size() > 0) {
                    dequeueFromLast.push(flipFrom.pop());
                }
            }

            if (!dequeueing) {
                while (enqueueingStacks.size() > 0) {
                    dequeueingStacks.push(enqueueingStacks.pop());
                }

                dequeueFrom = dequeueingStacks.pop();
                dequeueing = true;
            }

            if (dequeueFrom.size() == 0) {
                if (dequeueingStacks.size() > 0) {
                    dequeueFrom = dequeueingStacks.pop();
                } else {
                    dequeueFrom = dequeueFromLast;
                }
            }

            return dequeueFrom.pop();
        }
};

File: p5/solution_code/Python/Queue.py
from Stack import Stack

class Queue:
    def __init__(self):
        self.enqueueingStacks = Stack()
        self.dequeueingStacks = Stack()

        self.flipTo = Stack()
        self.flipFrom = Stack()
        self.enqueueTo = Stack()

        self.dequeueFrom = Stack()
        self.dequeueFromLast = Stack()

        self.stackExponent = 0
        self.dequeueing = False

    def enqueue(self, item):
        if self.flipFrom.size() > 0:
            self.flipTo.push(self.flipFrom.pop())

        self.enqueueTo.push(item)

        if self.enqueueTo.size() == 2**self.stackExponent:
            self.enqueueingStacks.push(self.flipTo)
            self.flipTo = self.flipFrom
            self.flipFrom = self.enqueueTo
            self.enqueueTo = Stack()
            self.stackExponent += 1

    def dequeue(self):
        for _ in range(4):
            if self.enqueueTo.size() > 0:
                self.dequeueFromLast.push(self.enqueueTo.pop())
            elif self.flipTo.size() > 0:
                self.flipFrom.push(self.flipTo.pop())
            elif self.flipFrom.size() > 0:
                self.dequeueFromLast.push(self.flipFrom.pop())

        if not self.dequeueing:
            while self.enqueueingStacks.size() > 0:
                self.dequeueingStacks.push(self.enqueueingStacks.pop())

            self.dequeueFrom = self.dequeueingStacks.pop()
            self.dequeueing = True

        if self.dequeueFrom.size() == 0:
            if self.dequeueingStacks.size() > 0:
                self.dequeueFrom = self.dequeueingStacks.pop()
            else:
                self.dequeueFrom = self.dequeueFromLast

        return self.dequeueFrom.pop()
